### DISCUSSION TITLE
Custom Relation Extraction component does not improve

### DISCUSSION THREAD

**Discussion by sighalt** [explosion/spaCy#13576] [Help: Coding & Implementations] [answered]
I have been trying for some time to create my own Relation Extraction Component based on your tutorial. However, my problem is that the component just doesn't want to learn and I don't know what the problem could be.
In the first step, I trained my own NER component, which also works well. On this basis, I train the Relation Extraction Component with the following configuration.

During training, however, Recall, Precision and F-Score remain at 0.0 and do not improve. I have added additional debugging output to check if `get_instances` returns matching candidates (‘get_instances PRF’ in the output), which it does.
However, it never recognises even one relation (‘Relation PRF’ in the output).

Does anyone have any ideas what else I could try?  I have already tried to increase the learning rate and adjust the score_weights without success.

Configuration:

```
[paths]
train = null
dev = null
raw = null
init_tok2vec = null

[system]
seed = 342
gpu_allocator = "pytorch"

[nlp]
lang = "de"
pipeline = ["transformer", "ner", "relation_extractor"]
disabled = []
before_creation = null
after_creation = null
after_pipeline_creation = null
tokenizer = {"@tokenizers":"spacy.Tokenizer.v1"}
batch_size = 24

[components]

[components.relation_extractor]
factory = "relation_extractor"
threshold = 0.5

[components.relation_extractor.model]
@architectures = "rel_model.v1"

[components.relation_extractor.model.create_instance_tensor]
@architectures = "rel_instance_tensor.v1"

[components.relation_extractor.model.create_instance_tensor.tok2vec]
@architectures = "spacy-transformers.TransformerListener.v1"
grad_factor = 1.0

[components.relation_extractor.model.create_instance_tensor.tok2vec.pooling]
@layers = "reduce_mean.v1"

[components.relation_extractor.model.create_instance_tensor.pooling]
@layers = "reduce_mean.v1"

[components.relation_extractor.model.create_instance_tensor.get_instances]
@misc = "rel_instance_generator.v1"
max_length = 100

[components.relation_extractor.model.classification_layer]
@architectures = "rel_classification_layer.v1"
nI = null
nO = null


[components.ner]
source = "data/spacy-ner.model/model-last/"
component = "ner"

[components.transformer]
source = "data/spacy-ner.model/model-last/"
component = "transformer"

[initialize]

[initialize.components]

[corpora]

[corpora.dev]
@readers = "spacy.Corpus.v1"
path = ${paths.dev}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[corpora.train]
@readers = "spacy.Corpus.v1"
path = ${paths.train}
max_length = 0
gold_preproc = false
limit = 0
augmenter = null

[training]
seed = ${system.seed}
gpu_allocator = ${system.gpu_allocator}
dropout = 0.1
accumulate_gradient = 1
patience = 1600000
max_epochs = 0
max_steps = 10000
eval_frequency = 500
frozen_components = ["transformer","ner"]
dev_corpus = "corpora.dev"
train_corpus = "corpora.train"
before_to_disk = null
logger = {"@loggers":"spacy.ConsoleLogger.v1"}
annotating_components = ["transformer","ner"]


[training.batcher]
@batchers = "spacy.batch_by_padded.v1"
discard_oversize = true
size = 200
buffer = 96
get_length = null


[training.optimizer]
@optimizers = "Adam.v1"
beta1 = 0.9
beta2 = 0.999
L2_is_weight_decay = true
L2 = 0.01
grad_clip = 1.0
use_averages = false
eps = 0.00000001
learn_rate = 0.001

[training.score_weights]
rel_micro_p = 0.0
rel_micro_r = 0.0
rel_micro_f = 1.0
```

Output of the training:

```
ℹ Saving to output directory: data/spacy-rel.model
ℹ Using GPU: 0

=========================== Initializing pipeline ===========================
*redacted*/venv/lib/python3.12/site-packages/transformers/utils/generic.py:441: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
*redacted*/venv/lib/python3.12/site-packages/transformers/utils/generic.py:309: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
*redacted*/venv/lib/python3.12/site-packages/transformers/utils/generic.py:309: UserWarning: torch.utils._pytree._register_pytree_node is deprecated. Please use torch.utils._pytree.register_pytree_node instead.
  _torch_pytree._register_pytree_node(
✔ Initialized pipeline

============================= Training pipeline =============================
ℹ Pipeline: ['transformer', 'ner', 'relation_extractor']
ℹ Frozen components: ['transformer', 'ner']
ℹ Set annotations on update for: ['transformer', 'ner']
ℹ Initial learn rate: 0.001
E    #       LOSS RELAT...  ENTS_F  ENTS_P  ENTS_R  REL_MICRO_P  REL_MICRO_R  REL_MICRO_F  SCORE 
---  ------  -------------  ------  ------  ------  -----------  -----------  -----------  ------
ℹ Relation PRF: {'tp': 21, 'tn': 112406, 'fp': 12005, 'fn': 224}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
  0       0           1.08   96.82   96.26   97.38         0.17         8.57         0.34    0.49
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
  2     500           9.04   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
  4    1000           6.67   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
  7    1500           6.65   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
  9    2000           6.49   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 11    2500           6.88   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 14    3000           7.09   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 16    3500           7.54   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 18    4000           6.36   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 21    4500           7.30   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 23    5000           6.63   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 25    5500           6.71   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 28    6000           6.69   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 30    6500           6.47   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 32    7000           7.19   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 35    7500           6.98   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 37    8000           6.95   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 39    8500           6.68   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 42    9000           6.54   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 44    9500           6.50   96.82   96.26   97.38         0.00         0.00         0.00    0.48
ℹ Relation PRF: {'tp': 0, 'tn': 124411, 'fp': 0, 'fn': 245}
ℹ get_instances PRF: {'tp': 1843, 'fp': 18933, 'fn': 12656}
 46   10000           7.32   96.82   96.26   97.38         0.00         0.00         0.00    0.48
✔ Saved pipeline to output directory
data/spacy-rel.model/model-last
```

**Comment by sighalt** [comment, marked as answer]
After hours of debugging where the error came from, I discovered a bug in the conversion of my label-studio data into DocBin format.
My question has thus become superfluous.
